### DISCUSSION TITLE
General USB-Reader support (No need for HID)

### DISCUSSION THREAD

**Discussion by powertomato** [MiczFlor/RPi-Jukebox-RFID#2160] [Ideas]
Hi I just implemented a PoC for a `Reader.py` for generic USB NFC readers, no need for them to have a keyboard HID driver.
It's based on [nfcpy](https://github.com/nfcpy/nfcpy) and my initial tests are successful. Possibly it could be compatible with the HID driver readers, but I cannot test it.

It uses libusb, which is a user space USB driver. Because of that additional work is required to get it running: a udev rule to make it available to non-root users and a modprobe.conf to blacklist the regular kernel drivers from loading.
Both of those things depend on the actual device. The udev rule requires the vender and product ID of the reader, while the blacklist entry requires knowlege which module is corresponding to the device.

I'm curious if the community is interested in this, and if I should create a pull request? The two required config files would make it difficult to create an automatic install script. Except we crowd-source compatible readers and maintain a list.

Here are my changes, other than those files I used a regular v2.4 installation.

Edit: I forgot, I installed 3 additional python packages: 
`sudo apt install python3-pyserial`
`sudo apt install python3-libusb1`
`sudo python -m pip install nfcpy`

`/etc/udev/rules.d/50-usb-nfc-rule.rules`:
```
SUBSYSTEM=="usb", ATTRS{idVendor}=="072f", ATTRS{idProduct}=="2200", MODE="0666"
```

`/etc/modprobe.d/disable_pn533_usb.conf`:
```
install pn533_usb /bin/true
```

`/home/pi/RPi-Jukebox-RFID/scripts/Reader.py`:
```.py
import os.path
import sys

import nfc
from nfc.clf import RemoteTarget

class NamedDevice(object):
    name = None

    def __init__(self, name):
        self.name = name

def get_devices():
    return [NamedDevice('nfcpy')]


class Reader:
    def __init__(self):
        self.reader = self
        self.clf = nfc.ContactlessFrontend('usb')

    def readCard(self):
        while True:
            target = self.clf.sense(RemoteTarget('106A'), RemoteTarget('106B'), RemoteTarget('212F'), interval=0.1, iterations=0xFFFFFFFFFFFFFFFF)
            if not target:
                continue

            tag = nfc.tag.activate(self.clf, target)
            if not tag:
                continue
            
            id = ''
            for char in tag.identifier:
                id += '%02X' % char

            return id

    def cleanup(self):
        self.clf.close()
```

**Comment by s-martin** [comment]
Hi, thanks for sharing your additions! 🎉

Of course we are always interested in improvements and using more readers.

I suggest you have a look at the documentation for the new Phoniebox version 3.x (branch `future3/develop`), which is implemented purely in Python and has a more thorough architecture.
We do not plan to implement new features in 2.x.

Start here: https://github.com/MiczFlor/RPi-Jukebox-RFID/tree/future3/develop/documentation

Readers are described here (including a template for new readers): https://github.com/MiczFlor/RPi-Jukebox-RFID/tree/future3/develop/documentation/developers/rfid

Developer info is here: https://github.com/MiczFlor/RPi-Jukebox-RFID/tree/future3/develop/documentation/developers

You can also meet us in the chat: https://matrix.to/#/#phoniebox_community:matrix.org

**Comment by powertomato** [comment]
Thanks for your reply. I'm on a rock64 and diet-pi distribution. I could not get v3 to run, partly because of the CPU check, so that would have been the third changed variable. Since this is supposed to be a Christmas present for my daughter I stuck with the release version for now.

Glad to hear there is interest in the reader module I wrote. Hopefully I will have some time around the holidays to port it to v3.

**Reply by s-martin** [reply]
You could disable the CPU/32bit check here ;-) https://github.com/MiczFlor/RPi-Jukebox-RFID/blob/24423c36d26c4b9e2656870d9fd5a27c22eb6e65/installation/install-jukebox.sh#L97

**Reply by powertomato** [reply]
I know, but when I did it all went to hell 😅
The log was full with "tee: /dev/fd/3: Permission denied"
The script didn't write anything on the screen

And I think it concluded with a compilation error of a pip module, which looked it could be related to the processor check

**Comment by pabera** [comment]
If you don't mind me asking.. Why have you chosen a [Rock64](https://www.pine64.org/devices/single-board-computers/rock64/) and diet-pi alongside it? The hardware seems to be pretty strong for the Jukebox.

**Reply by powertomato** [reply]
I have a couple of those lying around. I bought them as a "Iconikal Rockchip"  and the boards were sourced from a failed security monitor called [Recon Sentinel](https://hackaday.com/2020/11/16/teardown-recon-sentinel/). It's not clear if the SBCs had manufacturing problems or if they simply overproduced. They sold for 10€ each and came with SD card, power supply, a two-line ASCII display and a pretty ugly case.

As for diet-pi. There simply is not much alternatives to it. Armbian was dead slow with about 2 minutes boot time and diet-pi seemed like a good start for a headless application. It works great and boots in 30s, half of which is Wifi setup. And I didn't even optimize the boot yet.
And it's debian based, so much closer to rasbian than say Arch, so I figured my chances to get it running are high.

**Comment by s-martin** [comment]
Reference: implemented in #2190
